core/trino-main/src/main/java/io/trino/operator/RowReference.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+/**
+ * Reference to a row.
+ * <p>
+ * Note: RowReference gives us the ability to defer row ID generation (which can be expensive in tight loops).
+ */
+public interface RowReference
+{
+    /**
+     * Compares the referenced row to the specified row ID using the provided RowIdComparisonStrategy.
+     */
+    int compareTo(RowIdComparisonStrategy strategy, long rowId);
```

Comment:
it's odd that `compareTo` takes `RowIdComparisonStrategy` as argument since `RowIdComparisonStrategy` requires both `long leftRowId, long rightRowId`. As I understand `RowReference` might not have row id yet, so how such rows can be compared?

Reply:
How the comparison occurs is left as an implementation detail of the RowReference. For example: for all the caller cares about, the RowReference *could* be allocating a rowID, making a comparison, and deallocating a rowID. In practice, there are some internal details that allow it to make the comparison with the underlying cursor representation. See the comment in https://github.com/trinodb/trino/pull/6518#discussion_r562886614 for more details.